static/js/message_store.js

```diff
@@ -128,19 +129,21 @@ function add_message_metadata(message) {
 
     // Add new people involved in this message to the people list
     _.each(involved_people, function (person) {
-        // Do the hasOwnProperty() call via the prototype to avoid problems
-        // with keys like "hasOwnProperty"
-        if (! people.get_by_email(person.email)) {
-            people.add(person);
-        }
-
-        if (people.get_by_email(person.email).full_name !== person.full_name) {
-            people.reify(person);
-        }
+        if (!person.unknown_local_echo_user) {
+            if (! people.get_by_email(person.email)) {
+                people.add({
+                    email: person.email,
+                    user_id: person.user_id || person.id,
+                    full_name: person.full_name,
+                    is_admin: person.is_realm_admin || false,
+                    is_bot: person.is_bot || false
```

Comment:
Yeah, I see what you mean.  The `id` vs. `user_id` thing is something we could address by doing an API version bump for the relevant endpoints, which we should plan to do but maybe not quite yet (would like to change other things if we're doing a version bump).  

Doing fetching kinda sucks since it's extra network traffic; there's work to batch; and work to make it async.  

For sending the data with messages, there's 3 places where we send messages to clients:
* edited message edit (already has all the people from the original message, so I think no work here?).
* new message event
* get_old_messages

I think the backend does know when sending a message which users are referenced and which ones a client "should" have, I think it could just send down the people dicts for either all user IDs or just those the client "shouldn't" have batched in another part of the get_old_messages response (would be a lot more compressible than sending the display recipient / sender data each time).  We could put the backend changes under an option similar to `use_markdown` so that we don't impact any existing API clients.  

So I guess I'd be tempted to move in the direction of having the backend use the fact that it knows what the client should be getting; for being able to support changing email addresses.

I guess overall conclusion is that I think this is fine for now.